### PR TITLE
Dependency installer script

### PR DESCRIPTION
**Pull request by habibayassin**

### CODE REVIEW

File: etc/DependencyInstaller.sh
Comment (vvbandeira) [inline etc/DependencyInstaller.sh]: Let's use the pre-built distribution for each OS instead of building from source.

File: etc/DependencyInstaller.sh
Comment (vvbandeira) [inline etc/DependencyInstaller.sh]: I think `brew install python` provides `pip`, so there's no need to install it again.

File: etc/DependencyInstaller.sh
Comment (vvbandeira) [inline etc/DependencyInstaller.sh:91]: You will likely need to run `brew link --force libomp`.

File: etc/DependencyInstaller.sh
Comment (vvbandeira) [inline etc/DependencyInstaller.sh]: Add `-U` just in case it is already installed.

File: .github/workflows/github-actions-cron-test-installer.yml
Comment (vvbandeira) [inline .github/workflows/github-actions-cron-test-installer.yml]: doesn't `submodule: recursive` works?

File: .github/workflows/github-actions-cron-test-installer.yml
Comment (vvbandeira) [inline .github/workflows/github-actions-cron-test-installer.yml]: Why do we need this?

File: .github/workflows/github-actions-cron-test-installer.yml
Comment (vvbandeira) [inline .github/workflows/github-actions-cron-test-installer.yml]:
```suggestion
  testInstaller-macos:
```

File: .github/workflows/github-actions-cron-test-installer.yml
Comment (vvbandeira) [inline .github/workflows/github-actions-cron-test-installer.yml]:
```suggestion
      - name: Run installer
```

File: etc/DependencyInstaller.sh
Comment (vvbandeira) [inline etc/DependencyInstaller.sh:39]: Which tool depends on ruby?

File: etc/DependencyInstaller.sh
Comment (vvbandeira) [inline etc/DependencyInstaller.sh]: `pip` packages can be bundled in a common function for all OSes.

**Review by vvbandeira — CHANGES_REQUESTED** [review]

File: .github/workflows/github-actions-cron-test-installer.yml
Comment (habibayassin) [reply .github/workflows/github-actions-cron-test-installer.yml]: because the script can't run even with sudo as it creates a directory in /opt for or tools

File: etc/DependencyInstaller.sh
Comment (habibayassin) [reply etc/DependencyInstaller.sh:39]: klayout I read in the documentation it needs it as a dependency https://www.klayout.de/build.html

File: .github/workflows/github-actions-cron-test-installer.yml
Comment (habibayassin) [reply .github/workflows/github-actions-cron-test-installer.yml]: it causes this error when it runs inside a container: https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/actions/runs/3780081571/jobs/6425873507

**Comment by habibayassin** [comment]
@vvbandeira Notes on docker helper:
- the images are named as flow-<os>-<target>
- it doesn't have test functionality like the one in OR
- it uses the defaults in OR installation and build no options to change them

File: etc/DependencyInstaller.sh
Comment (vvbandeira) [inline etc/DependencyInstaller.sh]: `-dev` has been deprecated

File: etc/DependencyInstaller.sh
Comment (vvbandeira) [inline etc/DependencyInstaller.sh]: remove since we are also providing macos packages

File: etc/DependencyInstaller.sh
Comment (vvbandeira) [inline etc/DependencyInstaller.sh]: lets remove line 128 to have the same behavior as OR installer

File: flow/Makefile
Comment (vvbandeira) [inline flow/Makefile]: let's do a python file with some comments, this oneliner got a bit more complex than I expected

File: .github/workflows/github-actions-cron-test-installer.yml
Comment (vvbandeira) [reply .github/workflows/github-actions-cron-test-installer.yml]: The latest dependency installer should not install anything on /opt on macOS machines. https://github.com/The-OpenROAD-Project/OpenROAD/blob/master/etc/DependencyInstaller.sh#L135

File: .github/workflows/github-actions-cron-test-installer.yml
Comment (vvbandeira) [inline .github/workflows/github-actions-cron-test-installer.yml:62]: We need two jobs for macOS. One using [GH runners](https://github.com/The-OpenROAD-Project/OpenROAD/blob/master/.github/workflows/github-actions-macos.yml#L10-L11) and one with the [self-hosted runner](https://github.com/The-OpenROAD-Project/OpenROAD/blob/master/.github/workflows/github-actions-macos.yml#L41-L42).

File: build_openroad.sh
Comment (vvbandeira) [inline build_openroad.sh]: I was under the impression that by sourcing the `enable` file, this would not be required.

File: etc/DependencyInstaller.sh
Comment (vvbandeira) [inline etc/DependencyInstaller.sh]: since we are moving to python script this can be removed

File: etc/DependencyInstaller.sh
Comment (vvbandeira) [inline etc/DependencyInstaller.sh]: also remove gawk here

**Review by vvbandeira — CHANGES_REQUESTED** [review]

**Comment by oharboe** [comment]
@vvbandeira this is tough to review. It would be good to have a paragraph or two about what is being done here and why.

**Comment by vvbandeira** [comment]
@oharboe, the DependencyInstaller.sh and DockerHelper.sh will work in a similar way as their OpenROAD counterparts. The installer will handle any dependency independent of OS and if it is local or docker installs. The helper will generate docker images for any supported OS (initially centos and ubuntu, in the future RHEL, OpenSUSE, and Debian).

**Comment by vijayank88** [comment]
@vvbandeira  @habibayassin 
Can you update MacOS or-tools http link.? It's missing from the repo.
```
dependents or dependents with broken linkage!
--2023-01-16 09:09:07--  https://github.com/google/or-tools/releases/download/v9.5/or-tools_x86_64_MacOsX-13.0.1_cpp_v9.5.2237.tar.gz
Resolving github.com (github.com)... 192.30.255.112
Connecting to github.com (github.com)|192.30.255.112|:443... connected.
HTTP request sent, awaiting response... 404 Not Found
2023-01-16 09:09:07 ERROR 404: Not Found.
```

**Comment by habibayassin** [comment]
> @vvbandeira @habibayassin Can you update MacOS or-tools http link.? It's missing from the repo.
> 
> ```
> dependents or dependents with broken linkage!
> --2023-01-16 09:09:07--  https://github.com/google/or-tools/releases/download/v9.5/or-tools_x86_64_MacOsX-13.0.1_cpp_v9.5.2237.tar.gz
> Resolving github.com (github.com)... 192.30.255.112
> Connecting to github.com (github.com)|192.30.255.112|:443... connected.
> HTTP request sent, awaiting response... 404 Not Found
> 2023-01-16 09:09:07 ERROR 404: Not Found.
> ```

@vijayank88 it will be resolved once [this](https://github.com/The-OpenROAD-Project/OpenROAD/pull/2750) is merged

**Comment by vvbandeira** [comment]
> @vvbandeira @habibayassin Can you update MacOS or-tools http link.? It's missing from the repo.
> 
> ```
> dependents or dependents with broken linkage!
> --2023-01-16 09:09:07--  https://github.com/google/or-tools/releases/download/v9.5/or-tools_x86_64_MacOsX-13.0.1_cpp_v9.5.2237.tar.gz
> Resolving github.com (github.com)... 192.30.255.112
> Connecting to github.com (github.com)|192.30.255.112|:443... connected.
> HTTP request sent, awaiting response... 404 Not Found
> 2023-01-16 09:09:07 ERROR 404: Not Found.
> ```

@vijayank88 should be fixed now.

**Comment by vijayank88** [comment]
OpenROAD submodule need to be updated here and hope this PR will pass to resolve macos test installer error.

File: etc/DependencyInstaller.sh
Comment (vvbandeira) [inline etc/DependencyInstaller.sh]: Similar to OR, common with sudo/root should not use `--user` as it can install things in `/root/.local`.

File: etc/DockerHelper.sh
Comment (vvbandeira) [inline etc/DockerHelper.sh]: Let's use the same naming convention as OR with the minor version included.

File: etc/DockerHelper.sh
Comment (vvbandeira) [inline etc/DockerHelper.sh]: Same

File: etc/DockerHelper.sh
Comment (vvbandeira) [inline etc/DockerHelper.sh]: for-loop

File: etc/DockerHelper.sh
Comment (vvbandeira) [inline etc/DockerHelper.sh]: add support for macos

File: etc/DependencyInstaller.sh
Comment (vvbandeira) [inline etc/DependencyInstaller.sh:194]: Check for root/sudo usage. brew + sudo is not recommended.

File: etc/DependencyInstaller.sh
Comment (vvbandeira) [inline etc/DependencyInstaller.sh]: Let's define `klayoutVersion` and any other dependency version at the top of the script file and use for all installer methods/pkg managers.

File: etc/DependencyInstaller.sh
Comment (vvbandeira) [inline etc/DependencyInstaller.sh]: Add support for the new flags, `-base`, `-common`, etc. The ORFS installer should receive the same subset of arguments and forward them to OR installer.

**Review by vvbandeira — CHANGES_REQUESTED** [review]

File: etc/DependencyInstaller.sh
Comment (habibayassin) [reply etc/DependencyInstaller.sh:194]: it doesn't work with sudo I tried it in one of the GHA it says it is not recommended and exits

Comment (vvbandeira) [reply etc/DependencyInstaller.sh:194]: Right, so should we add a check in the script and let the user know that they cannot use `sudo` + installer on macOS? Because the error message from `brew` might not be clear enough, it will say that brew should not be used with `sudo`, not the installer.

**Comment by vijayank88** [comment]
@vvbandeira Can you please check this PR status?

**Comment by vvbandeira** [comment]
@vijayank88 there are problems in CI with the latest OR commit. We need to fix it before merging. I am looking into it.

**Comment by vijayank88** [comment]
@habibayassin 
Can you please merge latest commits from ORFS master and check PR is passing or not?
I'm waiting for this PR to merge in master to install in MacOS.

**Comment by habibayassin** [comment]
> @habibayassin Can you please merge latest commits from ORFS master and check PR is passing or not? I'm waiting for this PR to merge in master to install in MacOS.

@vijayank88 it is pending a fix to be merged on master that will resolve the failing Jenkins jobs.

File: docker/Dockerfile.dev
Comment (vvbandeira) [inline docker/Dockerfile.dev]:
```suggestion
COPY tools/OpenROAD/etc/DependencyInstaller.sh /tmp/tools/OpenROAD/etc/DependencyInstaller.sh

RUN /tmp/DependencyInstaller.sh \
      && rm -f /tmp/DependencyInstaller.sh \
      && rm -f /tmp/tools/OpenROAD/etc/DependencyInstaller.sh
```
should we add `/tmp`?

**Review by vvbandeira — COMMENTED** [review]
@habibayassin this looks almost ready to merge, could you just check why the docker build stage is failing to find `or-tools`?

**Comment by habibayassin** [comment]
> @habibayassin this looks almost ready to merge, could you just check why the docker build stage is failing to find `or-tools`?

@vvbandeira why aren't we using `./etc/DockerHelper.sh create -target=builder -os=<os-name>` for building the docker images? shouldn't we remove the old way of building the images and update the docs?

**Comment by vvbandeira** [comment]
> > @habibayassin this looks almost ready to merge, could you just check why the docker build stage is failing to find `or-tools`?
> 
> @vvbandeira why aren't we using `./etc/DockerHelper.sh create -target=builder -os=<os-name>` for building the docker images? shouldn't we remove the old way of building the images and update the docs?

I think keeping the `build_openroad.sh` makes sense, but you can change how this scripts builds the docker image internally.

**Review by vvbandeira — APPROVED** [review]

**Pull request merged**
